src/dataflow/src/render/sources.rs

```diff
@@ -621,13 +622,32 @@ where
                     }
                 };
 
-                if let Some(errors) = errors {
+                if let Some(errors) = errors_collection {
                     error_collections.push(errors);
                 }
 
-                (stream, capability)
+                (ok_collection, capability)
             };
 
+            // This is the end of the "source pipeline", in as far as, this
+            // collection is expected to be definite, and return the same data
+            // at the same timestamps (modulo compaction) across restarts.
+            //
+            // Beyond this point, there may be source-instance specific computation
```

Comment:
I don't think this is true, for UPSERT, we inject an `as_of` and linear operator (which includes temporal filters) into the upsert operator already. (I have thoughts on whether this is a problem or not, though, advancing the timestamp via an as_of or temporal filters does not downgrade capabilities or advance the actual timely timestamp, at least in the current impl. I'll ask about this in our internal slack as well...)

Reply:
Right ok so I think theres two things here:

1. There isn't currently a clean cut boundary between STORAGE and COMPUTE for all kinds of sources whereby we can say "once you cross this line all collections are definite". This pr is trying to help suss out where that boundary is / should be so theres definitely bits here that are aspirational. I can amend the documentation to reflect that.

2. IIUC: theres two classes of MFPs - those that are on the SOURCE, vs those that are on the VIEW. MFPs on the source itself can be pushed as far down into storage as we want, and we can we can use that MFP to decide what data to store or not. MFPs on the view cannot be used to decide what to store, but they can be passed to storage and storage can optionally decide "oh this mfp only needs this subset of columns. i will only retrieve that."

I'm not sure if today we even expose syntax for mfps tied to the source. @frankmcsherry does the above sound right?

Reply:
1. If you want to work on this boundary more intentionally, there are specific plans for what that boundary should reflect. In particular, there is meant to be prescriptive ways of determining where the boundary is, vs descriptive (refactoring and looking). Probably no harm in doing the looking, but I think when rubber meets the road, thing will need to be brought into line with the prescriptive approach.

2. IIRC you can specify a hunk of `MIR` attached to a source, as part of "source orthogonality", and leaving that in place for sources seems great (tell the source about as much about what *must* happen as possible). I can't find it in the code at the moment though. `SourceDesc` has the view-specific MFPs, currently as a `LinearOperator` though we'd be happy to improve that to MFPs; these cannot be pushed all the way to the source definition as they are not shared, but STORAGE can accept them an in principle push them all the way up to that (e.g., use times in MFPs to only load subsets of the data).